src/main/java/com/minecolonies/coremod/colony/buildings/AbstractCitizenAssignable.java

```diff
@@ -0,0 +1,255 @@
+     */
+    public AbstractCitizenAssignable(final BlockPos pos, final Colony colony)
+    {
+        super(pos);
+        this.colony = colony;
+    }
+
+    @Override
+    public void readFromNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.readFromNBT(compound);
+        assignedCitizen.clear();
+    }
+
+    @Override
+    public void writeToNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.writeToNBT(compound);
+    }
+
+    /**
+     * Returns the colony of the building.
+     *
+     * @return {@link com.minecolonies.coremod.colony.Colony} of the current object.
+     */
+    @NotNull
+    public Colony getColony()
+    {
+        return colony;
+    }
+
+    /**
+     * Method to do things when a block is destroyed.
+     */
+    public void onDestroyed()
+    {
+        if (hasAssignedCitizen())
+        {
+            // EntityCitizen will detect the workplace is gone and fix up it's
+            // Entity properly
+            assignedCitizen.clear();
+        }
+    }
+
+    /**
+     * On tick of the server.
+     *
+     * @param event {@link net.minecraftforge.fml.common.gameevent.TickEvent.ServerTickEvent}
+     */
+    public void onServerTick(final TickEvent.ServerTickEvent event)
+    {
+        // Can be overridden by other buildings.
+    }
+
+    /**
+     * On tick of the world.
+     *
+     * @param event {@link net.minecraftforge.fml.common.gameevent.TickEvent.WorldTickEvent}
+     */
+    public void onWorldTick(final TickEvent.WorldTickEvent event)
+    {
+        // Can be overridden by other buildings.
+    }
+
+    //------------------------- Starting Assigned Citizen handling -------------------------//
+
+    /**
+     * Get the main workerbuildings of the building (the first in the list).
+     *
+     * @return the matching CitizenData.
+     */
+    public CitizenData getMainCitizen()
+    {
+        if (assignedCitizen.isEmpty())
+        {
+            return null;
+        }
+        return assignedCitizen.get(0);
+    }
+
+    /**
+     * Returns the workerbuildings of the current building.
+     *
+     * @return {@link CitizenData} of the current building
+     */
+    public List<CitizenData> getAssignedCitizen()
+    {
+        return new ArrayList<>(assignedCitizen);
+    }
+
+    /**
+     * Method to remove a citizen.
+     *
+     * @param citizen Citizen to be removed.
+     */
+    public void removeCitizen(final CitizenData citizen)
+    {
+        if (isCitizenAssigned(citizen))
+        {
+            assignedCitizen.remove(citizen);
+            citizen.setWorkBuilding(null);
+            markDirty();
+        }
+    }
+
+    /**
+     * Returns if the {@link CitizenData} is the same as the workerbuildings.
+     *
+     * @param citizen {@link CitizenData} you want to compare
+     * @return true if same citizen, otherwise false
+     */
+    public boolean isCitizenAssigned(final CitizenData citizen)
+    {
+        return assignedCitizen.contains(citizen);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') This call to "contains()" may be a performance hot spot if the collection is large. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)